### ISSUE TITLE
Read_Parquet too slow between versions 1.* and 2.*

### ISSUE THREAD

**Issue by CrashLaker** [dask/dask#6376]
Hi all,

I'd like to report a weird slow `read_parquet` behavior between dask versions.
On all examples I'll be using:
* CentOS 7.7
* Python 3.6
* Fastparquet 0.4.0 (Installed with pip)

Dask versions were also installed with pip.
`pip3 install dask=={version} dask[dataframe]=={version}`

The I'm using is:
```python
filters = [('x', '>=', time_from), ('x', '<', time_to)]
ddf = dd.read_parquet('path', columns=['x', 'y'], filters=filters, engine='fastparquet')
s = datetime.datetime.now().timestamp()
df = ddf.compute()
e = datetime.datetime.now().timestamp()
print('took', e-s, 'secs')
```

**Test 1** <--- slow
dask==2.20.0
fastparquet==0.4.0
the output is:
`took 62.25287580490112 secs`

**Test  2** <--- slow
dask==2.18.1
fastparquet==0.4.0
output:
`took 62.15111708641052 secs`

**Test 3** <--- fast
dask==1.2.2
fastparquet==0.4.0
output:
`took 1.4967741966247559 secs`

The data schema is as shown below:
`df.head()`
|    | x                   |    y |
|---:|:--------------------|-----:|
|  0 | 2020-06-24 13:00:05 | 1003 |
|  1 | 2020-06-24 13:00:10 | 1083 |
|  2 | 2020-06-24 13:00:15 | 1247 |
|  3 | 2020-06-24 13:00:20 | 1173 |
|  4 | 2020-06-24 13:00:25 | 1260 |

`df.dtypes`
```
x    datetime64[ns]
y           float64
dtype: object
```

**What happened**:
read_parquet function is slow on recent dask versions > 1min

**What you expected to happen**:
faster read_parquet function.
dask=1.2.2 took 1sec

Regards,
C.

**Comment by gforsyth** [comment]
Thanks for reporting @CrashLaker -- can you provide a reproducible example?  https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

It will make it much easier for us to help debug this.

**Comment by CrashLaker** [comment]
Hi @gforsyth 
thank you for your reply.
Also thank you for the super useful [post](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Although I'm not sure what's really happening I'm only able to reproduce it If have the full dataset which is about 50MB size.

Also I tried to randomize the data before sharing and found another strange issue that is:
If I read and save the dataset it reduces considerably in size and the query time goes back to normal.
**real dataset**: ./test2
**copied dataset**: ./test3

Could you spot any possible issues?
```python3
real_db = "./test2"
to_db = "./test3"

print(dd.__version__) # '2.20.0'
print(fastparquet.__version__) # '0.4.0'

ddf = dd.read_parquet(real_db) # ./test2
df = ddf.compute() # takes 2min ish

fastparquet.write(
    filename=to_db, # ./test3
    data=df,
    compression='GZIP',
    file_scheme='hive',
    partition_on=['year', 'month', 'day'],
    write_index=False
)

ddf = dd.read_parquet(to_db) # read ./test3
df = ddf.compute() # takes 900ms
```

Though the dataset folder is way small than the original one:
```bash
du -sh ./test2 # 50MB
du -sh ./test3 # 15MB
```

About the data I did a quick len check and also checked it's head content
```python
len(df1) == len(df2) # True
len(df1) == len(df1.drop_duplicates) # True
df1.head()
df2.head()
```
Everything seems fine.

Regards,
C.

**Comment by gforsyth** [comment]
Hey @CrashLaker -- it's hard to say for sure, but a couple of thoughts:

If the original parquet dataset is made up of very many files and doesn't have a `_metadata` file, that could slow down processing (but I'm surprised by the 2 minute mark for a 50MB dataset).  How was it generated in the first place?

If it's partitioned as your example suggests (when you write the dataset back out) and you only have a few entries per calendar date, I could see that creating a LOT of partitions.

What do you see for

```
ddf.npartitions
```

when you load `real_db`?

**Comment by CrashLaker** [comment]
Every hour I load data from the source and dump it to the parquet folder using this code where
`toadd_new` is a pandas DataFrame with the new data.

```python3
fastparquet.write(**{
    "filename": real_db,
    "data": toadd_new,
    "compression": 'GZIP',
    "file_scheme": 'hive',
    "partition_on": partition_on, # ['year', 'month', 'day']
    "write_index": False,
    **append
})
```

the data is very granular as it's a timeseries database with 5/5s interval data like the example below:
|    | ds                  |   A |   B |   C | ... |
|---:|:--------------------|-------:|----------:|----------:| ----:|
|  0 | 2020-06-24 00:00:00 |   2572 |      1116 |      1456 | ..|
|  0 | 2020-06-24 00:00:05 |  15377 |      9375 |      6002 | ..|
|  1 | 2020-06-24 00:00:10 |  19022 |      9795 |      9227 | ..|
|  2 | 2020-06-24 00:00:15 |  23096 |     14829 |      8267 | ..|
|  3 | 2020-06-24 00:00:20 |  26605 |     22569 |      4036 | ..|

```python3
pritn(dask.__version__) # 2.20.0

df = dd.read_parquet(real_db)
df.npartitions # 438

df = dd.read_parquet(copied_db)
df.npartitions # 14
```

`real_db` has 438 partitions. Could this be an issue?
I'm not really that familiar using Dask, parquet, etc.

thank you for your help.
C.

**Comment by mrocklin** [comment]
Thank you for raising an issue @CrashLaker .  Unfortunately it is hard for us to see what is going on with your computation.  There are a few things things that you can do to help (any of them would be useful):

1.  Provide a minimal reproducible example. Something that we can copy-paste into a Python terminal to reproduce your situation.  Probably you will have to create random data for this.
2.  Provide a performance report.  See https://docs.dask.org/en/latest/diagnostics-distributed.html#capture-diagnostics and maybe also the video in that section.
3.  Use `git bisect` on the dask repository to find out when the performance regression occurred.

Without those we are just guessing, and this is an inefficient use of time.

**Comment by CrashLaker** [comment]
Hi @mrocklin 
thank you for your reply.
> Without those we are just guessing, and this is an inefficient use of time.
agree with you.

I'm not able to create provide a reproducible example as when I randomize the data and save it to another location
the problem disappears.
Is there a way to edit the data and save it using the same structure as I was read?

Anyways the video on 2. was really insightful as I could dump the performance report. Could you please take a look on?

This is the performance dump on the `read_db` path (taking ~90seconds):
https://rawcdn.githack.com/CrashLaker/plainrepo/eb2e896d2f5b03a21c4dbe9be22e2a1d204a2033/dask-report/dask-report-real-db.html

Also this is the performance dump of the `copied_db`. (taking ~1second). both have the same data. I just loaded from `read_db` and dumped it on another path as described earlier.
https://rawcdn.githack.com/CrashLaker/plainrepo/eb2e896d2f5b03a21c4dbe9be22e2a1d204a2033/dask-report/dask-report-copied-db.html

regards,
C.

**Comment by mrocklin** [comment]
It looks like you're spending most of your time in fastparquet, and in
there compiling numba code.

Short term you might try using read_parquet(engine="pyarrow") to see if
there is some difference.

Long term cc'ing @martindurant in case something is off with numba
compilation.

**Comment by martindurant** [comment]
- I am surprised that the workers are spending any time reading thrift; is there a _metadata file? Unfortunately, if there isn't, then reading each file's footer is indeed required, since it contains the bytes offsets for the data (and it's an all-of-nothing read, you can't just get the offsets).
- waiting for a lock on the numba compile is weird, I haven't seen that before. Presumably each worker is wanting to compile the *same* thing, and this should happen when the task arrives to the worker, and only once. Maybe it's just showing up as the overhead grows compared to the small data size. Numba is able to cache compiled functions, but this has never worked for the "jit-classes" used by fastparquet. I am not aware of any progress on this.

**Comment by CrashLaker** [comment]
Hi @martindurant 
Thank you all for the support.

By saying metadata file you mean the `_metadata` file it creates at the root of the folder?
```
$ tree
.
├── _common_metadata
├── _metadata
└── year=2020
    ├── month=6
    │   ├── day=20
    │   │   ├── part.0.parquet
    │   │   ├── part.1.parquet
    │   │   ├── part.2.parquet
    │   │   ├── part.3.parquet
    │   │   ├── part.4.parquet
    │   │   └── part.5.parquet
    │   ├── day=21
    │   │   ├── part.10.parquet
    │   │   ├── part.11.parquet
    │   │   ├── part.12.parquet

```

By using pyarrow==0.17.1 dask==2.20.0
the time goes down to 12s
https://rawcdn.githack.com/CrashLaker/plainrepo/4d59a487ac2269fd0b479f7e857e06cd6387b22e/dask-report/dask-report-dask-pyarrow.html

The problem is that using fastparquet==0.4.0 and dask==1.2.2 the elapsed time goes down to 1s
But I can't seem to get performance_report to work on dask=-1.2.2

```python
ImportError: cannot import name 'performance_report' from 'dask.distributed' (/opt/conda/lib/python3.7/site-packages/dask/distributed.py)
```

**Comment by mrocklin** [comment]
Now it looks like we're mostly bound on Pandas memory manipulation.  I
recommend trying with more processes and fewer threads.  Right now it looks
like you have one process with eight threads.  Maybe try 2-4 or 4-2?  If
you are using LocalCluster then see the threads_per_process keyword.

**Comment by CrashLaker** [comment]
Hi @mrocklin 
Though I was expecting it to run faster single process/threaded.
Like dask==1.2.2 (**1second**)

The elapsed times for the threaded ones haven't changed.

Workers = 2 threads_per_worker = 4 (**12secs**)
dask==2.20.0 / pyarrow==0.17.1
```python
from dask.distributed import Client, LocalCluster
cluster = LocalCluster(n_workers=2, threads_per_worker=4)
client = Client(cluster)
```
https://rawcdn.githack.com/CrashLaker/plainrepo/9c8b3fee494c2089388ca67820c5eaa05a40dd6b/dask-report/dask-report-dask-pyarrow-cluster-p2.t4.html

Workers=4 threads_per_worker = 2(**10secs**)
dask==2.20.0 / pyarrow==0.17.1
```python
from dask.distributed import Client, LocalCluster
cluster = LocalCluster(n_workers=4, threads_per_worker=2)
client = Client(cluster)
```
https://rawcdn.githack.com/CrashLaker/plainrepo/e82cd37bf452681ff5682172235fb215ceeab3a8/dask-report/dask-report-dask-pyarrow-cluster-p4.t2.html

**Comment by mrocklin** [comment]
It looks like about half of the time is taken up in opening files.
Again @martindurant might find this interesting.

Other than that everything looks like it's running at top-speed to me.  I'm
not sure how else to help without knowing more about the problem.  This is
where it would be useful to have a minimal reproducible example.

**Comment by martindurant** [comment]
Agreed, if you can produce code I can run, I'm happy to have a look. You could do this by writing many small files yourself (with random data), or perhaps writing from a dask dataframe after rechunk to a small number of rows per partition.

@rjzamora , you may find this discussion interesting in the light of https://github.com/dask/dask/pull/6346

**Comment by CrashLaker** [comment]
Thank you for the advice @martindurant !
Could reproduce it by writing a dummy code.

dask==2.20.0 ran in (1min1s)
dask==1.2.2 ran in (12s)

env
python3.6
centos7.7
dask==2.20.0
fastparquet==0.4.0

The code below took `1min 6s` to run on my machine.
I creates a folder ./test006 and appends random data to it
```python3
import fastparquet
import dask
import dask.dataframe as dd
import datetime
import random
import pandas as pd
import os

folder_save = './test006'
start = datetime.datetime(2020,6,1)
colsize = 8
for y in range(400):  # <----------------------- took 1min 6s
    row = []
    for i in range(720):
        start = start + datetime.timedelta(seconds=5)
        row.append([start, *[random.randint(1,1000) for i in range(colsize)]]) 
    df = pd.DataFrame(row)
    df.columns = ['ds', *[chr(65+i) for i in range(colsize)]]
    df['year'] = df['ds'].apply(lambda x: x.timetuple().tm_year)
    df['month'] = df['ds'].apply(lambda x: x.timetuple().tm_mon)
    df['day'] = df['ds'].apply(lambda x: x.timetuple().tm_mday)
    append = {'append': True} if os.path.exists(folder_save+'/_metadata') else {}
    fastparquet.write(**{
        "filename": folder_save,
        "data": df,
        "compression": 'GZIP',
        "file_scheme": 'hive',
        "partition_on": ['year', 'month', 'day'],
        "write_index": False,
        **append
    })
```

to read
```python3
import datetime
import dask
import dask.dataframe as dd

s = datetime.datetime.now().timestamp()
ddf = dd.read_parquet('./test006')
df = ddf.compute()
e = datetime.datetime.now().timestamp()
print('took', e-s, 'secs')  # <-------------------------- took 1min 1s dask==2.20.0
#                                                             took 12s dask==1.2.2
```

it took `1min 1s` to read on my machine

Regards,
C.

**Comment by martindurant** [comment]
@mrocklin FYI on opening files, the following diff changes the open time for me from 83us to 70us. This is perhaps worth doing, but clearly not the crux!

```diff
--- a/fsspec/implementations/local.py
+++ b/fsspec/implementations/local.py
@@ -2,6 +2,7 @@ import datetime
 import io
 import os
 import shutil
+import stat
 import posixpath
 import re
 import tempfile
@@ -58,12 +59,12 @@ class LocalFileSystem(AbstractFileSystem):
         path = self._strip_protocol(path)
         out = os.stat(path, follow_symlinks=False)
         dest = False
-        if os.path.islink(path):
+        if stat.S_ISLNK(out.st_mode):
             t = "link"
             dest = os.readlink(path)
-        elif os.path.isdir(path):
+        elif stat.S_ISDIR(out.st_mode):
             t = "directory"
-        elif os.path.isfile(path):
+        elif stat.S_ISREG(out.st_mode):
             t = "file"
         else:
             t = "other"
@@ -153,11 +154,14 @@ class LocalFileSystem(AbstractFileSystem):
         return True


+matcher = re.compile("/[A-Za-z]:")
+
+
 def make_path_posix(path, sep=os.sep):
     """ Make path generic """
     if isinstance(path, (list, set, tuple)):
         return type(path)(make_path_posix(p) for p in path)
-    if re.match("/[A-Za-z]:", path):
+    if matcher.match(path):
         # for windows file URI like "file:///C:/folder/file"
         # or "file:///C:\\dir\\file"
         path = path[1:]
@@ -165,7 +169,7 @@ def make_path_posix(path, sep=os.sep):
         # special case for windows UNC/DFS-style paths, do nothing,
         # just flip the slashes around (case below does not work!)
         return path.replace("\\", "/")
-    if re.match("[A-Za-z]:", path):
+    if matcher.match(path):
         # windows full path like "C:\\local\\path"
```

**Comment by martindurant** [comment]
(^ that of course on my laptop SSD - may be more important on other FSs, which could even have network latency, but I would think ought to be OS buffered)

@CrashLaker , I will try to find some time to run your benchmark.

**Comment by mrocklin** [comment]
Yeah, personally I would pass on that change until we can find that it
affects serious workloads.

**Comment by martindurant** [comment]
I find that fastparquet directly is significantly faster than via dask:
```
In [16]: %time pf.to_pandas()  # direct into single dataframe
Wall time: 615 ms

In [18]: %time pandas.concat(list(pf.iter_row_groups())) # roughly what dask does
Wall time: 2.49 s

In [21]: %time ddf.compute() # dask, threads
Wall time: 37.4 s

In [22]: %prun ddf.compute(scheduler='sync')
         17450324 function calls (17175174 primitive calls) in 37.899 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     4160   23.273    0.006   29.425    0.007 {built-in method thrift.protocol.fastbinary.decode_compact}
  1557504    2.564    0.000    2.564    0.000 ttypes.py:1261(__init__)
  1557504    1.850    0.000    1.850    0.000 ttypes.py:296(__init__)
  1557504    1.135    0.000    1.135    0.000 ttypes.py:1360(__init__)
   347360    0.862    0.000    0.862    0.000 {method 'findall' of 're.Pattern' objects}
      416    0.834    0.002    0.872    0.002 api.py:490(_dtypes)
181792/8320    0.425    0.000    1.308    0.000 util.py:376(unique_everseen)
  1557504    0.396    0.000    0.396    0.000 ttypes.py:1594(__init__)
  1557920    0.345    0.000    0.345    0.000 api.py:181(<genexpr>)
   178063    0.303    0.000    0.303    0.000 {method 'split' of 'str' objects}
```

The profile shows that *all* the time is spent in parsing thrift. There should be **none** of this at all, since all the metadata was already loaded before compute (takes 64ms). For some reason, we are re-reding the metadata of every data file, and in this case each file contains so little actual data, that this dominates.

@rjzamora , do you happen to immediately know why this is happening in the presence of a _metadata file? In FastParquetEngine.read_metadata, we seem to end up in case (3) at L369, as a means of reducing the size of the graph - in this case it seems misguided.

**Comment by rjzamora** [comment]
> @rjzamora , do you happen to immediately know why this is happening in the presence of a _metadata file? In FastParquetEngine.read_metadata, we seem to end up in case (3) at L369, as a means of reducing the size of the graph - in this case it seems misguided.

I believe we had users experiencing issues with large graph sizes, and over-prioritized the size of the graph over the overhead of needing to re-read metadata.  This is likely just lazy code on my part.  However, the best way to avoid the need for us to balance these pain points is to communicate information about partitioning in a different way (and avoid rereading _metadata at all costs).  Would it be more efficeint to communicate only the path, row-group index, and partitioning information to the `read_partition` method?  The read task would then need to create a new `ParquetFile` for the target path, and then use `ParquetFile.read_row_group` to read a specific row-group.  If the dataset is partitioned, we would need a way to handle that without re-reading _metadata.

**Comment by martindurant** [comment]
> Would it be more efficeint to communicate only the path, row-group index, and partitioning information

The information absolutely needed to read a give row group is: the path, the schema, partition info, and offsets of the columns. The first three are small and the schema is the same for all partitions, so it only needs to appear in the graph once. The offsets are encoded in the row-group blob and duplicated in the footer of each file, and all together in the metadata. I'd argue that the full set for all partitions must be small, but the fastparquet interface expects this info to be in a bunch of thrift structures - if we aren't to rewrite. Previously we did:
```
fp.row_groups = [row-group-of-interest]
task = read(fp, ...)
```
which made the object smaller per task, but potentially lots of similar tasks. Perhaps this should be

```
pf.row_groups = []
task = read(pf, row-group-of-interest, ...)
```
so we include the duplicated part once, and the differences only once.

There was also some code to descend into the row-group object to remove unnecessary things like the column stats, which maybe helps a little
```
In [14]: len(pickle.dumps(rg))
Out[14]: 1042

In [15]: for col in rg.columns:
    ...:     col.meta_data.encoding_stats = None
    ...:     col.meta_data.statistics = None
    ...:

In [16]: len(pickle.dumps(rg))
Out[16]: 745
```
(also improves pickle speed 25% in this case)

Note that pickle for fastparquet *also* uses thrift, which turns out faster than dealing with the deeply linked set of classes, at least with the older version of pickle from whenever I tested it.

**Comment by martindurant** [comment]
(just tested, thrift is still faster than pure pickle to roundtrip ser/de, but only by a whisker)

**Comment by CrashLaker** [comment]
Hi all.

Is there any workaround to this besides using a newer 1.2.2 version of it?
Is noticed the queries are slowly getting longer.
for instance:
```python
filters = [
	('ds', '>=', pd.Timestamp(2020,6,1),
	('ds', '<', pd.Timestamp(2020,7,1)
]
ddf = dd.read_parquet(dbfolder, columns=['ds', 'A'], filters=filters)
df = ddf.compute()
```

**Comment by martindurant** [comment]
We are working on it, please keep pinging this issue

**Comment by CrashLaker** [comment]
Hey guys,
any news on this?

**Comment by rjzamora** [comment]
Thanks for pinging @CrashLaker - We added some improvements to `ArrowEngine` that *should* help when you use `engine="pyarrow"`.  I haven't had a chance to attack the `FastParquetEngine` code yet.

**Comment by CrashLaker** [comment]
Hi @rjzamora,

thank you for your reply.
i'm trying to stick to fastparquet as an engine because later i'm planning to run this on lambda
and I'm aware there some issues with pyarrow due to its size being too big https://github.com/apache/arrow/issues/4216

maybe it has been [solved](https://issues.apache.org/jira/browse/ARROW-6776) will try it later

**Comment by jsignell** [comment]
@CrashLaker it seems like the pyarrow issue you referenced is closed. I suspect that moving from fastparquet to pyarrow is going to be the best path forward.

**Comment by CrashLaker** [comment]
Hi @jsignell 
thank you for your reply.

i haven't looked in depth to [it](https://github.com/apache/arrow/issues/4216) but afaik it's solved only by rebuilding it using fewer features and therefore resulting in a lighter package size.
Doable but not really ideal I believe.
That's why I wanned to solve things with fastparquet.

**Comment by CrashLaker** [comment]
Hi all,

I've been dumping data to parquet hive folder since I started this thread (jul 2020).
The process is: Scrape data from API from time to time then dump it to parquet.
which is pretty much like the script [above](https://github.com/dask/dask/issues/6376#issuecomment-656306726)
Question is: Is this process where I'm to append to the hive folder multiple times not recommended?

The problem is that this is taking longer and longer each time.
As it is saved to a hive scheme I can see the folders/files being organized as:
```
year=2020/
    month=12/
        day=1/
        day=2/
        ...
year=2021/
    month=1/
        day=1/
        day=2/
```

Not knowing anything about the underlying process I'd naively think that if I want to query all data from December 2020 it would then only look into the `year=2020/month=12` folder.

My dataset is not even large. It has over 400MB in size and over 10.000 partitions `ddf.npartitions`.

Querying it using:
* fastparquet 0.4.0 dask==2.20.0 pandas 1.1.5
    * took: 2 hr 51 min 1 sec
* fastparquet==0.4.0 dask==1.2.2 pandas 0.24
    * took: 3min 44sec
* pyarrow==2.0.0 dask==2.20.0
    * took:  4min 15sec

I believe that even it taking > 1min to run is a lot to my understand.

Regards,
C.

**Comment by martindurant** [comment]
Would you mind trying with `gather_statistics=False`?

**Comment by rjzamora** [comment]
I really appreciate that you are taking the time to share your experiences here @CrashLaker .  It has proved challenging to produce a `read_parquet` solution that is optimal for all three engines ("fastparquet", "pyarrow-legacy", "pyarrow-dataset") and all use cases (partitioning, filtering, scale, etc.).

>Question is: Is this process where I'm to append to the hive folder multiple times not recommended?

I think it is a common/reasonable workflow.  However, it does produce many small files, so reading the entire dataset can become very inefficient.  If you only want to read back a subset of data, passing the `filters` argument to read parquet **should** make the query much more efficient (especially for the "pyarrow-dataset" engine).

**General comments on pyarrow vs fastparquet engines**

I have not personally focused much on fastparquet recently, but I do know that, in order to avoid passing a large object for every task in the task graph, every task needs to re-read the global _metadata file to generate the `ParquetFile` object needed to produce the correct partition. **This is obviously far from ideal**, because you are opening and parsing the _metadata file 10,000 times for a 10,000-partition read (in addition to reading the actual data).  The pyarrow engine is faster, because the API makes it easier to read a row-group list directly from an isolated file.

Note that, in order to limit the size of the graph (and to enable the move to [`Blockwise`-based IO in the near future](https://github.com/dask/dask/pull/7042)), we cannot pass a pre-initialized `ParquetFile` to every task in the graph. We **need** to be able to produce an output partition from a list of simple information (like the path, row-group list, and hive-partitioning information).

I will think more about ways to improve the "fastparquet"-engine performance, and generally keep your script in mind in #7042 and #7066

**Comment by martindurant** [comment]
> Note that, in order to limit the size of the graph (and to enable the move to Blockwise-based IO in the near future), we cannot pass a pre-initialized ParquetFile to every task in the graph. We need to be able to produce an output partition from a list of simple information (like the path, row-group list, and hive-partitioning information).

I must have missed this change. Yes, it is totally possible to pass a ParquetFile with *no* row groups, so it would be the same for every task, and an argument which is the row group in question per task. Previously, indeed it was a modified ParquetFile for each task which, in hindsight, was probably not the most efficient thing to do even then.

**Comment by rjzamora** [comment]
>I must have missed this change. Yes, it is totally possible to pass a ParquetFile with no row groups, so it would be the same for every task, and an argument which is the row group in question per task. Previously, indeed it was a modified ParquetFile for each task which, in hindsight, was probably not the most efficient thing to do even then.

So, are you saying we can pass a stripped down `ParquetFile` object to every task (or embed it in the io_function for the `Blockwise` implementation), and then just include the path and row-groups as args to every task?  This seems good to me if the stripped-down `ParquerFile` object is small enough, and we can still call `pf.read_row_group_file`.

**Comment by martindurant** [comment]
We would need a small function to reattach the row-group to the ParquetFile before calling `read_row_group_file` but yes. That is, of course, *if* we have all the row-groups up front (i.e., _metadata file or full scan) - otherwise pass the paths as before and defer file footer reading to the workers.

**Comment by CrashLaker** [comment]
Hi thank you all for your help. Times are getting lower now.

> Would you mind trying with gather_statistics=False?
hi @martindurant 

this is strange `filters` seems to be getting ignored when disabling `gather_statistics=False`

tests with pyarrow==3.0.0 (fast) - now 18secs
dask==2.20.0
fastparquet==0.5.0
pandas==1.1.5
pyarrow==3.0.0

**took 18secs**
```python
folder_save = "./db"

s = datetime.datetime.now().timestamp()
time_from = '2020-12-01'
time_to = '2021-01-01'
time_from = pd.Timestamp(2020,12,1).tz_localize('UTC')
time_to = pd.Timestamp(2021,1,1).tz_localize('UTC')
filters = [('ds', '>=', time_from), ('ds', '<', time_to)]

ddf = dd.read_parquet(folder_save, filters=filters, engine='pyarrow')

print("#partitions", ddf.npartitions) # <---- 1179
df = ddf.compute()
print("num elems", len(df)) # <-------- 8443
print(df.head())  # <--- prints okay from 1st Dec 2020
print(df.tail())   # <----- to 31st Dec 2020
```

**took 72seconds**
```python3
folder_save = "./db"

s = datetime.datetime.now().timestamp()
time_from = pd.Timestamp(2020,12,1).tz_localize('UTC')
time_to = pd.Timestamp(2021,1,1).tz_localize('UTC')
filters = [('ds', '>=', time_from), ('ds', '<', time_to)]

ddf = dd.read_parquet(folder_save, filters=filters, engine='pyarrow', gather_statistics=False)

print("#partitions", ddf.npartitions) # <---- 11522
df = ddf.compute()
print("num elems", len(df)) # <---- 63873
print(df.head()) # <---- wrong from 2020 June
print(df.tail()) # <----    to 2021 Feb
```

for comparison pyspark taking around **30secs**
spark-3.0.1-bin-hadoop2.7
pyspark==3.0.1

```python
import datetime
from pyspark import SparkContext
from pyspark.sql import SparkSession

spark = SparkSession.builder \
            .master('local') \
            .appName('myappname') \
            .getOrCreate()

db = "./db/"

time_from = '2020-12-01'
time_to = '2021-01-01'

s = datetime.datetime.now()
df1 = spark.read.parquet(db)
df = df1.where((df1['ds'] >= time_from) & (df1['ds'] < time_to))
df2 = df.toPandas()
df2 = df2.sort_values('ds')

e = datetime.datetime.now()
print(e-s)  # <------ 0:00:28s

print(df2.head(10))
print(df2.tail(10))
```

**Comment by martindurant** [comment]
> this is strange filters seems to be getting ignored when disabling gather_statistics=False

This is not surprising at all, but perhaps should be more clearly documented. Filtering works either by pathnames for directory-formatter parquet and/or column statistics within chunks. If you don't parse those statistics, because it takes too long as an up-front cost, then you can't know which chunks to ignore. It should be possible, though, to skip parsing a chunk once we get to it in the worker - but I don't know if that works yet or if it would be a big time saving (should work fine for the newer arrow datatset engine).

**Comment by rjzamora** [comment]
Note that `engine="pyarrow-dataset"` will perform row-wise filtering without requiring `gather_statistics=True` (because the filtering is done by the pyarrow Dataset API, rather than in dask itself).  We can certainly try to document filtering behavior better.

**Comment by rjzamora** [comment]
@CrashLaker - Curious if you tested with the latest "fastparquet" engine as well?  I would expect "pyarrow" and "pyarrow-dataset" to be a bit faster, but the fastparquet version *should* have improved lately.

**Comment by CrashLaker** [comment]
Oh yeah!
Now I'm liking these numbers. There's been so much improvement since i first created the issue.
congrats guys!

> @CrashLaker - Curious if you tested with the latest "fastparquet" engine as well? I would expect "pyarrow" and "pyarrow-dataset" to be a bit faster, but the fastparquet version should have improved lately.

latest dask
dask==2021.2.0
fastparquet==0.5.0
pandas==1.1.5
pyarrow==3.0.0

**engine='pyarrow'** took avg **13secs**

**engine='fastparquet'** too avg **11secs**

@rjzamora 
Using `engine='pyarrow-dataset'` gives me `segmentation fault` though

**Comment by martindurant** [comment]
> engine='pyarrow' took avg 13secs
> engine='fastparquet' too avg 11secs

Sometime fastparquet still wins :)

**Comment by rjzamora** [comment]
>Now I'm liking these numbers. There's been so much improvement since i first created the issue.

Woohoo

>Using engine='pyarrow-dataset' gives me segmentation fault though

Grrr- I will need to try out pyarrow-3.0...  Perhaps there is a bug.

**Comment by jrbourbeau** [comment]
Since https://github.com/dask/dask/pull/7200 has been merged is there anything else to do here? Or are we okay to close?

**Comment by CrashLaker** [comment]
Hi @jrbourbeau 

Is #7200 part of `dask==2021.3.0` already? I can't tell.
If yes then I'm afraid I'm still facing `segmentation fault`.

`pip freeze`
```
dask==2021.3.0
numpy==1.20.1
pandas==1.2.3
pyarrow==3.0.0
```

But still the main problem I was having when I first opened this issue **has been solved**.
So to me this can be closed now.
Again appreciate all your attention and hardworking :)

**Comment by jrbourbeau** [comment]
https://github.com/dask/dask/pull/7200 was included in the `2021.03.0` release

Thanks for following up here @CrashLaker -- feel free to open a new issue for the seg fault
